File: Leetcode/temp/0589_N-ary_Tree_Preorder_Traversal.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        travesal = []
        if not root:
            return []
        if root.val is not None:
            travesal.append(root.val)
        if root.children:
            roots = root.children
        else:
            return travesal
        roots = roots[::-1]
        while roots != []:
            r = roots.pop()
            travesal.append(r.val)
            if r.children:
                roots += r.children[::-1]
        return travesal

File: Leetcode/Java/0589_N-ary_Tree_Preorder_Traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public List<Integer> preorder_stack(Node root) {
        List<Integer> nodes = new LinkedList<>();
        Stack<Node> stack = new Stack<>();
        if (root == null) {
            return nodes;
        }
        stack.push(root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            nodes.add(node.val);
            List<Node> children = node.children;
            Collections.reverse(children);
            for (Node child: children) {
                stack.push(child);
            }
        }
        return nodes;
    }

    // recursive
    public List<Integer> preorder_recursive(Node root) {
        List<Integer> list = new LinkedList<>();
        helper(root, list);
        return list;
    }
    private void helper(Node root, List<Integer> list) {
        if (root == null) {
            return;
        }
        list.add(root.val);
        for (Node child: root.children) {
            helper(child, list);
        }
    }
}